tests/devices/default_qubit/test_default_qubit_native_mcm.py

```diff
@@ -60,52 +70,69 @@ def validate_samples(shots, results1, results2):
     Otherwise, fails if the sums of samples differ by more than ``20`` plus 20 percent.
     """
     if isinstance(shots, Sequence):
-        assert isinstance(results1, Sequence)
-        assert isinstance(results2, Sequence)
-        assert len(results1) == len(results2)
+        assert isinstance(results1, tuple)
+        assert isinstance(results2, tuple)
+        assert len(results1) == len(results2) == len(shots)
         for s, r1, r2 in zip(shots, results1, results2):
-            validate_samples(s, r1, r2)
-    else:
-        sh1, sh2 = results1.shape[0], results2.shape[0]
-        assert np.allclose(sh1, sh2, rtol=20, atol=0.2)
-        assert results1.ndim == results2.ndim
-        if results2.ndim > 1:
-            assert results1.shape[1] == results2.shape[1]
-        np.allclose(np.sum(results1), np.sum(results2), rtol=20, atol=0.2)
+            validate_samples(s, r1, r2, batch_size=batch_size)
+        return
+
+    if batch_size is not None:
+        assert isinstance(results1, Iterable)
+        assert isinstance(results2, Iterable)
+        assert len(results1) == len(results2) == batch_size
+        for r1, r2 in zip(results1, results2):
+            validate_samples(shots, r1, r2, batch_size=None)
+        return
+
+    sh1, sh2 = results1.shape[0], results2.shape[0]
+    assert np.allclose(sh1, sh2, rtol=20, atol=0.2)
+    assert results1.ndim == results2.ndim
+    if results2.ndim > 1:
+        assert results1.shape[1] == results2.shape[1]
+    np.allclose(np.sum(results1), np.sum(results2), rtol=20, atol=0.2)
 
 
-def validate_expval(shots, results1, results2):
+def validate_expval(shots, results1, results2, batch_size=None):
     """Compares two expval, probs or var.
 
     If the results are ``Sequence``s, validate the average of items.
 
     If ``shots is None``, validate using ``np.allclose``'s default parameters.
     Otherwise, fails if the results do not match within ``0.01`` plus 20 percent.
     """
-    if isinstance(results1, Sequence):
-        assert isinstance(results2, Sequence)
-        assert len(results1) == len(results2)
+    if isinstance(shots, Sequence):
+        assert isinstance(results1, tuple)
+        assert isinstance(results2, tuple)
+        assert len(results1) == len(results2) == len(shots)
         results1 = reduce(lambda x, y: x + y, results1) / len(results1)
         results2 = reduce(lambda x, y: x + y, results2) / len(results2)
-        validate_expval(shots, results1, results2)
+        validate_expval(sum(shots), results1, results2, batch_size=batch_size)
         return
+
     if shots is None:
         assert np.allclose(results1, results2)
```

Comment:
This supports batching?

Reply:
If `shots=None`, the device defaults to `defer_measurements`, which handles batching much more cleanly. Moreover, which expectation values, we don't have to worry about the shape of the output so we can just use `allclose` rather than iterating over the batch dim and validating.

Reply:
Might be missing the context, but this seems to work as intended:

```
dev = qml.device("default.qubit", shots=[10, 2, 3])

@qml.qnode(dev)
def f(x):
    for i in range(100):
        qml.RY(x, 0)
        m = qml.measure(0, postselect=1)
    return qml.expval(qml.PauliZ(0))

x = np.array([0.4, 0.6])
f(x)
```

```
(array([nan, nan]), array([nan, nan]), array([nan, nan]))
```